src/riak_kv_compile_tab.erl

```diff
@@ -71,35 +138,77 @@ is_compiling(BucketType) ->
         compiling_state() | notfound.
 get_state(BucketType) when is_binary(BucketType) ->
     case dets:lookup(?TABLE, BucketType) of
-        [{_,_,_,State}] ->
+        [{_,_,_,_,State}] ->
             State;
         [] ->
             notfound
     end.
 
 %%
+-spec get_compiled_ddl_version(BucketType :: binary()) ->
+    riak_ql_ddl_compiler:compiler_version() | notfound.
+get_compiled_ddl_version(BucketType) when is_binary(BucketType) ->
+    case dets:match(?TABLE, {BucketType,'$1','_','_','_'}) of
+        [[Version]] ->
+            Version;
+        [] ->
+            notfound
+    end.
+
+%%
 -spec get_ddl(BucketType :: binary()) ->
         term() | notfound.
 get_ddl(BucketType) when is_binary(BucketType) ->
     case dets:lookup(?TABLE, BucketType) of
-        [{_,DDL,_,_}] ->
+        [{_,_,DDL,_,_}] ->
             DDL;
         [] ->
             notfound
     end.
 
 %% Update the compilation state using the compiler pid as a key.
+%% Since it has an active Pid, it is assumed to have a DDL version already.
 -spec update_state(CompilerPid :: pid(), State :: compiling_state()) ->
-        ok | notfound.
+        ok | error | notfound.
 update_state(CompilerPid, State) when is_pid(CompilerPid),
                                        ?is_compiling_state(State) ->
-    case dets:match(?TABLE, {'$1','$2',CompilerPid,'_'}) of
-        [[BucketType, DDL]] ->
-            insert(BucketType, DDL, CompilerPid, State);
+    case dets:match(?TABLE, {'$1','$2','$3',CompilerPid,'_'}) of
+        [[BucketType, DDLVersion, DDL]] ->
+            insert(BucketType, DDLVersion, DDL, CompilerPid, State);
         [] ->
             notfound
     end.
 
+%% Mark any lingering compilations as being retried
+-spec mark_compiling_for_retry() -> ok.
+mark_compiling_for_retry() ->
+    CompilingPids = dets:match(?TABLE, {'_','_','_','$1',compiling}),
+    lists:foreach(fun([Pid]) -> update_state(Pid, retrying) end, CompilingPids).
+
+%% Get the list of records which need to be recompiled
+-spec get_ddl_records_needing_recompiling(DDLVersion :: riak_ql_ddl_compiler:compiler_version()) ->
+    [binary()].
+get_ddl_records_needing_recompiling(DDLVersion) ->
+    %% First find all tables with a version
+    MismatchedTables = dets:select(?TABLE, [{{'$1','$2','_','_',compiled},[{'/=','$2', DDLVersion}],['$$']}]),
+    RetryingTables = dets:match(?TABLE, {'$1','$2','_','_',retrying}),
+    Tables = [hd(X) || X <- MismatchedTables ++ RetryingTables],
+    lager:info("Recompile the DDL of these bucket types ~p", [Tables]),
+    Tables.
+
+%% TODO: Remove this once 1.2 support has been removed
+%% Convert all pre-1.3 records to include the DDL compiler version in the
+%% DETS table.  Simply recompile everything all the time since we don't
+%% know if this is a downgrade or not.
```

Comment:
Comment is incorrect. Rather, we don't know if this is an upgrade following a downgrade.

We could determine whether this was a downgrade from a later compiler version by comparing version fields in the 1.3 table.